workload/schemachange support user defined schemas in workload

jayshrivastava opened the pull request:
Adds support for user-defined schemas 
- create schema op
- drop schema op
- updates tables, enums, and views to be qualified with schema name prefixes

Resolves: #54408
Release note: None

jayshrivastava committed:
workload/schemachange: added createSchema operation

Release note: None

jayshrivastava committed:
workload/schemachange: added deleteSchema operation

Release note: None

jayshrivastava committed:
workload/schemachange: qualified all table names with schema names

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54889)
<!-- Reviewable:end -->

jayshrivastava committed:
workload/schemachange: qualified enums with schema names

Release note: None

jayshrivastava committed:
workload/schemachange: updated views to be qualified with schema names

Resolves: https://github.com/cockroachdb/cockroach/issues/54408

Release note: None

jayshrivastava commented on pkg/workload/schemachange/schemachange.go:
```diff
@@ -552,13 +556,13 @@ func (w *schemaChangeWorker) createSequence(tx *pgx.Tx) (string, error) {
 
 	stmt := rowenc.RandCreateTable(w.rng, "table", int(atomic.AddInt64(w.seqNum, 1)))
-	stmt.Table = tree.MakeUnqualifiedTableName(tree.Name(tableName))
+	stmt.Table = tree.MakeTableNameWithSchema((tree.Name)(w.database), (tree.Name)(schemaName), (tree.Name)(tableName))
```
I'm not sure if specifying the best db is the best thing to do here. It might be better to leave it blank until the workload supports multiple dbs.

jayshrivastava commented on pkg/workload/schemachange/schemachange.go:
```diff
@@ -1086,6 +1091,36 @@ func (w *schemaChangeWorker) randType(tx *pgx.Tx) (tree.ResolvableTypeReference,
+		return "", err
+	}
+
+	stmt := rowenc.MakeSchemaName(w.rng.Intn(2) == 0, schemaName, "root")
```
It may be better to leave authorization plank until the workload supports multiple roles.

ajwerner reviewed (commented):
> Reviewed 2 of 3 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54889)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava)

___
*[pkg/workload/schemachange/schemachange.go, line 1100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MILdOV5Zf98NyRwNL9X-r1-1100:-MIdOnkE5t6Cd1eFV7Vw:benef3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5d8a085cb901c94c9c8875615604bba06f7fd0b/pkg/workload/schemachange/schemachange.go#L1100)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

It may be better to leave authorization plank until the workload supports multiple roles.
</blockquote></details>

I agree completely. We should just leave a TODO to support authorization. Also worth noting that it is valid to not provide a name when you provide an authorization as it implies the schema of the same name.
___
*[pkg/workload/schemachange/schemachange.go, line 1110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MIdP8JG2XuvYWT9oXm_:-MIdP8JG2XuvYWT9oXma:b-i6s5qj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5d8a085cb901c94c9c8875615604bba06f7fd0b/pkg/workload/schemachange/schemachange.go#L1110)):*
> ```Go
> 	const q = `
>   SELECT *
>     FROM [information_schema.schemata]
> ```

This query is suspect to me. Firstly, I don't think those where brackets make sense. Secondly, if you do `SELECT *` I think it's going to fail when you only try to scan a single column. 


I think you want:

```
  SELECT schema_name
    FROM information_schema.schemata
   WHERE schema_name LIKE 'schema%' OR schema_name = 'public'
ORDER BY random()
   LIMIT 1;
```

I'm starting to think that we need to make it so that any errors coming from these schema introspection functions fail the workload completely. My read is that they currently do not. 
___
*[pkg/workload/schemachange/schemachange.go, line 565 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MILdOV5Zf98NyRwNL9X-r3-565:-MIdR7ec-y72P0Bzv3a1:b-nv7jpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ca209d03681163dea33fa264a892b7c4d7676eb/pkg/workload/schemachange/schemachange.go#L565)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

I'm not sure if specifying the best db is the best thing to do here. It might be better to leave it blank until the workload supports multiple dbs.
</blockquote></details>

This seems fine. If you'd prefer to not use the database name you could do:

```
tree.MakeTableNameFromPrefix(tree.ObjectPrefix{
    SchemaName:     tree.Name(schemaName),
    ExplicitSchema: true,
}, tree.Name(tableName))
```

Also nit: you don't need parents around the type name for the cast.
___
*[pkg/workload/schemachange/schemachange.go, line 584 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MIdS5OCEBz_NOYALBCO:-MIdS5OCEBz_NOYALBCP:bv0e5ro) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ca209d03681163dea33fa264a892b7c4d7676eb/pkg/workload/schemachange/schemachange.go#L584)):*
> ```Go
> 		return "", err
> 	}
> 	qualifiedTableName := fmt.Sprintf("%s.%s", schemaName, tableName)
> ```

On this whole commit I think it'd be better to have a helper here that can choose to omit public or something like that. I'd take it one step further and have `randTable` just return a string or a `tree.TableName` (which has a `String()` method)

The repetition in this commit isn't very valuable. 
___
*[pkg/workload/schemachange/schemachange.go, line 1077 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MIdT0VPBoXnygzGU4va:-MIdT0VPBoXnygzGU4vb:b-fb46a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff3a5dcfda335a0c32d7ce8af63c6c74659c9923/pkg/workload/schemachange/schemachange.go#L1077)):*
> ```Go
> }
> 
> func (w *schemaChangeWorker) randView(tx *pgx.Tx, pctExisting int) (string, string, error) {
> ```

same comment on the return types here, returning two strings makes the caller do a lot of work and makes the code harder to read as far as I can tell. what do you think?


<!-- Sent from Reviewable.io -->

jayshrivastava commented on pkg/workload/schemachange/schemachange.go:
```diff
@@ -1049,13 +1030,16 @@ ORDER BY random()
+func (w *schemaChangeWorker) randTable(tx *pgx.Tx, pctExisting int) (tree.TableName, error) {
 	if w.rng.Intn(100) >= pctExisting {
-		randSchema, err := w.randSchema(tx, pctExisting)
+		randSchema, err := w.randSchema(tx, 90)
```
pctExisting` will be a low value when `randTable` is called with the intention of creating a table, so a non existing table is returned most of the time. Before I added schemas, this means that a create table operation would mostly succeed because most of the time a non existing table name is returned. However, when you pass `pctExisting` to rand schema, the schema will most likely not exist. This would cause the create table operation to fail most of the time. To restore the original behaviour of normally succeeding, I thought 90 would be a good value here. 

The same thing applies to createEnum and createView.

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54889)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/workload/schemachange/schemachange.go, line 584 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MIdS5OCEBz_NOYALBCO:-MJ2IzsvCpcMBnlxbWAX:b10aaev) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ca209d03681163dea33fa264a892b7c4d7676eb/pkg/workload/schemachange/schemachange.go#L584)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

On this whole commit I think it'd be better to have a helper here that can choose to omit public or something like that. I'd take it one step further and have `randTable` just return a string or a `tree.TableName` (which has a `String()` method)

The repetition in this commit isn't very valuable. 
</blockquote></details>

Ok. I Changed the return type. 
___
*[pkg/workload/schemachange/schemachange.go, line 1077 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MIdT0VPBoXnygzGU4va:-MJ2InMf3mMV-cjiOAU1:b-5g6maw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff3a5dcfda335a0c32d7ce8af63c6c74659c9923/pkg/workload/schemachange/schemachange.go#L1077)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

same comment on the return types here, returning two strings makes the caller do a lot of work and makes the code harder to read as far as I can tell. what do you think?
</blockquote></details>

Ok. It makes sense to me. I just put out these changes. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Seems like you need a rebase. This is getting close. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-:-MJ95wH74PQzFH-ljvNf:bvjm1ax)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/workload/schemachange/schemachange.go, line 1035 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MILdOV5Zf98NyRwNL9X-r12-1035:-MJ94vABABHCxyl_l6T3:b4szgdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/53a592e1cc444ad05cd481eb9568effeb672f0de/pkg/workload/schemachange/schemachange.go#L1035)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

`pctExisting` will be a low value when `randTable` is called with the intention of creating a table, so a non existing table is returned most of the time. Before I added schemas, this means that a create table operation would mostly succeed because most of the time a non existing table name is returned. However, when you pass `pctExisting` to `randSchema`, the schema will most likely not exist. This would cause the create table operation to fail most of the time. To restore the original behaviour of normally succeeding, I thought 90 would be a good value here. 

The same thing applies to createEnum and createView.  
</blockquote></details>

interesting, I feel like if we're choosing to not return an existing name we know that no matter what, what we return here will not exist. The question at hand is whether the schema also does not exist. I don't know how to be principled about what to do here. Picking a random constant feels worse than just propagating `pctExisting`, no? 
___
*[pkg/workload/schemachange/schemachange.go, line 502 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94eUu5_pJdxMUppS9:-MJ94eUu5_pJdxMUppSA:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L502)):*
> ```Go
> 	}
> 	def.Nullable.Nullability = tree.Nullability(rand.Intn(1 + int(tree.SilentNull)))
> 	return fmt.Sprintf(`ALTER TABLE "%s" ADD COLUMN %s`, tableName.String(), tree.Serialize(def)), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 600 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94co65gT0mOTQ8Hdp:-MJ94co65gT0mOTQ8Hdq:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L600)):*
> ```Go
> 
> 	return fmt.Sprintf(`CREATE TABLE "%s" AS SELECT %s FROM "%s"`,
> 		tableName.String(), tree.Serialize(&names), destTableName.String()), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 640 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94bjzBm-v5qG70aBu:-MJ94bjzBm-v5qG70aBv:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L640)):*
> ```Go
> 		return "", err
> 	}
> 	return fmt.Sprintf(`ALTER TABLE "%s" DROP COLUMN "%s"`, tableName.String(), columnName), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 652 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94bTYBHqaz_8K-Wb0:-MJ94bTYBHqaz_8K-Wb1:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L652)):*
> ```Go
> 		return "", err
> 	}
> 	return fmt.Sprintf(`ALTER TABLE "%s" ALTER COLUMN "%s" DROP DEFAULT`, tableName.String(), columnName), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 664 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94Xy06x_L0KmgO8tw:-MJ94Xy06x_L0KmgO8tx:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L664)):*
> ```Go
> 		return "", err
> 	}
> 	return fmt.Sprintf(`ALTER TABLE "%s" ALTER COLUMN "%s" DROP NOT NULL`, tableName.String(), columnName), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 747 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94gMC6K9k2SkOyUwt:-MJ94gMC6K9k2SkOyUwu:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L747)):*
> ```Go
> 
> 	return fmt.Sprintf(`ALTER TABLE "%s" RENAME COLUMN "%s" TO "%s"`,
> 		tableName.String(), srcColumnName, destColumnName), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 767 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94h3W5XUxtoaU2xaN:-MJ94h3W5XUxtoaU2xaO:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L767)):*
> ```Go
> 
> 	return fmt.Sprintf(`ALTER TABLE "%s" RENAME CONSTRAINT "%s" TO "%s"`,
> 		tableName.String(), srcIndexName, destIndexName), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 795 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94hg06ro-vRNGdhjx:-MJ94hg06ro-vRNGdhjy:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L795)):*
> ```Go
> 	}
> 
> 	return fmt.Sprintf(`ALTER TABLE "%s" RENAME TO "%s"`, srcTableName.String(), destTableName.String()), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 809 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94iCv2vaTBGRC4di0:-MJ94iCv2vaTBGRC4di1:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L809)):*
> ```Go
> 	}
> 
> 	return fmt.Sprintf(`ALTER VIEW "%s" RENAME TO "%s"`, srcViewName.String(), destViewName.String()), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 827 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94jAB5Hi4EeyZWA4R:-MJ94jAB5Hi4EeyZWA4S:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L827)):*
> ```Go
> 		return "", err
> 	}
> 	return fmt.Sprintf(`ALTER TABLE "%s" ALTER COLUMN "%s" SET NOT NULL`, tableName.String(), columnName), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 844 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94jvd09qYZiuJtRFg:-MJ94jve6YFiq7mdN3dH:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L844)):*
> ```Go
> 	}
> 	return fmt.Sprintf(`ALTER TABLE "%s" ALTER COLUMN "%s" SET DATA TYPE %s`,
> 		tableName.String(), columnName, typ), nil
> ```

nit: the `%s` verb calls `String()`
___
*[pkg/workload/schemachange/schemachange.go, line 872 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94kgf173Wfb-mJzoQ:-MJ94kgf173Wfb-mJzoR:b-ykqcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L872)):*
> ```Go
> 	return fmt.Sprintf(
> 		`INSERT INTO "%s" (%s) VALUES %s`,
> 		tableName.String(),
> ```

nit: the `%s` verb calls `String()`


<!-- Sent from Reviewable.io -->

jayshrivastava committed:
workload/schemachange: update out of date usage comment

Updated the 'Example usage' comment at the top of the workload
to use the correct flag names and types

Release note: None

jayshrivastava committed:
workload/schemachange: remove quotations from table names in sql statements

This fixes a syntax error. Specifying `"public.table_name"` instead of
`public.table_name` will not recognize `public` as the schema name.

Release note: None

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54889)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/workload/schemachange/schemachange.go, line 1035 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MILdOV5Zf98NyRwNL9X-r12-1035:-MJDeYAV2P84VVnVreXg:btzdv1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/53a592e1cc444ad05cd481eb9568effeb672f0de/pkg/workload/schemachange/schemachange.go#L1035)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

interesting, I feel like if we're choosing to not return an existing name we know that no matter what, what we return here will not exist. The question at hand is whether the schema also does not exist. I don't know how to be principled about what to do here. Picking a random constant feels worse than just propagating `pctExisting`, no? 
</blockquote></details>

It is worth noting that we do use int literals in a few places, but that I plan to clean up all of them with my work related to error classification and control. 

I think `100 - pctExisting` might be the best. We mainly use `pctExisting` as a way to control errors. For example, we call `w.randTable(tx, 10)` in `createTable`. I feel that the intention is to have it fail roughly 10% of the time due to an existing table being given, otherwise succeed 90% of the time. My problem with propagating`pctExisting` is that we will actually fail most of the time in that 90% case by calling `w.randSchema(tx, 10))`. The problem is we do not distinguish between if the error case is when a name does exist (ie. the table name) and when it doesn't (the schema name). There are a lot of examples of this in the workload (ex. create a column on a table - in the non error case, the column name does not exist, but the table name does). It makes sense to use two different `pctExisting`s here to achieve some overall error rate. 

Overall, I think the idea of `pctExisting` needs to be refactored into something called an `errorRate`. We should also configure it as a property in `schemaChangeWorker` rather than int literals. Then, we need to provide an abstraction that generates the pctExisting to use based on if the error case is having a name exist or not. I plan to accomplish this during my next project. 
___
*[pkg/workload/schemachange/schemachange.go, line 600 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94co65gT0mOTQ8Hdp:-MJDjX0d98PErYzGHVJF:b85ikg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L600)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: the `%s` verb calls `String()`
</blockquote></details>

It seems like this happens: `CREATE TABLE {{table1647 { public %!s(bool=false) %!s(bool=true)}}} AS SELECT col676_6, col676_3, col676_1 FROM {{table675 { public %!s(bool=false) %!s(bool=true)}}}` it also causes CI to fail due to linting. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
still needs a rebase against master, no? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-:-MJDnCo-5YwMGWb9rwOl:b7ysrpw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/workload/schemachange/schemachange.go, line 1035 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MILdOV5Zf98NyRwNL9X-r12-1035:-MJDn0215-2kZldEuqjn:b-kodbp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/53a592e1cc444ad05cd481eb9568effeb672f0de/pkg/workload/schemachange/schemachange.go#L1035)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

It is worth noting that we do use int literals in a few places, but that I plan to clean up all of them with my work related to error classification and control. 

I think `100 - pctExisting` might be the best. We mainly use `pctExisting` as a way to control errors. For example, we call `w.randTable(tx, 10)` in `createTable`. I feel that the intention is to have it fail roughly 10% of the time due to an existing table being given, otherwise succeed 90% of the time. My problem with propagating`pctExisting` is that we will actually fail most of the time in that 90% case by calling `w.randSchema(tx, 10))`. The problem is we do not distinguish between if the error case is when a name does exist (ie. the table name) and when it doesn't (the schema name). There are a lot of examples of this in the workload (ex. create a column on a table - in the non error case, the column name does not exist, but the table name does). It makes sense to use two different `pctExisting`s here to achieve some overall error rate. 

Overall, I think the idea of `pctExisting` needs to be refactored into something called an `errorRate`. We should also configure it as a property in `schemaChangeWorker` rather than int literals. Then, we need to provide an abstraction that generates the pctExisting to use based on if the error case is having a name exist or not. I plan to accomplish this during my next project. 
</blockquote></details>

sgtm. 
___
*[pkg/workload/schemachange/schemachange.go, line 600 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94co65gT0mOTQ8Hdp:-MJDmjnLA4gvILyo_Ypz:bgckna9) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L600)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

It seems like this happens: `CREATE TABLE {{table1647 { public %!s(bool=false) %!s(bool=true)}}} AS SELECT col676_6, col676_3, col676_1 FROM {{table675 { public %!s(bool=false) %!s(bool=true)}}}` it also causes CI to fail due to linting. 
</blockquote></details>

okay, what's going on here is that the string method is on the pointer. What if we changed the return type to be the pointer rather than the struct?


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
Yup. I'll rebase momentarily. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-:-MJDrAD92rdM-_j22u1t:b-rsigg4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/workload/schemachange/schemachange.go, line 600 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94co65gT0mOTQ8Hdp:-MJDqAqi0qlr3q2MPM3e:bsglmoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L600)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

okay, what's going on here is that the string method is on the pointer. What if we changed the return type to be the pointer rather than the struct?
</blockquote></details>

Because you cannot take the address of a fcn in Go, we would have to change all the return statements in `randTable` and `randView`   to look like this.
```
tableName := tree.MakeTableNameFromPrefix(tree.ObjectNamePrefix{
	SchemaName:     tree.Name(randSchema),
	ExplicitSchema: true,
}, tree.Name(fmt.Sprintf("table%d", atomic.AddInt64(w.seqNum, 1)))), nil
return tableName
```

of alternatively, we can update the `Sprintf`s to take the address.
fmt.Sprintf(`CREATE TABLE "%s" AS SELECT %s FROM %s`, &destTableName, tree.Serialize(&names), &tableName)


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54889)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/workload/schemachange/schemachange.go, line 600 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-MJ94co65gT0mOTQ8Hdp:-MJDyP3C6wB9mzKlmhuB:b-hlbh9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/017c5a826dd519ca45e26c90a086b02667f3fc2c/pkg/workload/schemachange/schemachange.go#L600)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

Because you cannot take the address of a fcn in Go, we would have to change all the return statements in `randTable` and `randView`   to look like this.
```
tableName := tree.MakeTableNameFromPrefix(tree.ObjectNamePrefix{
	SchemaName:     tree.Name(randSchema),
	ExplicitSchema: true,
}, tree.Name(fmt.Sprintf("table%d", atomic.AddInt64(w.seqNum, 1)))), nil
return tableName
```

of alternatively, we can update the `Sprintf`s to take the address.
fmt.Sprintf(`CREATE TABLE "%s" AS SELECT %s FROM %s`, &destTableName, tree.Serialize(&names), &tableName)
</blockquote></details>

I'll go with the first option. It'll keep the changes in 1 spot for each `rand` fcn.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54889#-:-MJSY9T68AloXVMmAze3:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)



<!-- Sent from Reviewable.io -->

jayshrivastava commented:
bors r+
